src/client/gameannouncer.py

```diff
@@ -1,61 +1,60 @@
-from PyQt5.QtCore import QTimer
 from model.game import GameState
-
-from fa import maps
+import chat.friendtracker
 
 
 class GameAnnouncer:
-    ANNOUNCE_DELAY_SECS = 35
 
-    def __init__(self, gameset, me, colors, client):
-        self._gameset = gameset
+    def __init__(self, playerset, me, colors, client):
         self._me = me
         self._colors = colors
         self._client = client
 
-        self._gameset.newLobby.connect(self._announce_hosting)
-        self._gameset.newLiveReplay.connect(self._announce_replay)
+        self._friends_event_tracker = chat.friendtracker.build_friends_tracker(me, playerset)
+        self._friends_event_tracker.friendEvent.connect(self._friend_event)
 
         self.announce_games = True
         self.announce_replays = True
-        self._delayed_host_list = []
-
-    def _is_friend_host(self, game):
-        return (game.host_player is not None
-                and self._me.isFriend(game.host_player.id))
-
-    def _announce_hosting(self, game):
-        if not self._is_friend_host(game) or not self.announce_games:
-            return
-        announce_delay = QTimer()
-        announce_delay.setSingleShot(True)
-        announce_delay.setInterval(self.ANNOUNCE_DELAY_SECS * 1000)
-        announce_delay.timeout.connect(self._delayed_announce_hosting)
-        announce_delay.start()
-        self._delayed_host_list.append((announce_delay, game))
-
-    def _delayed_announce_hosting(self):
-        timer, game = self._delayed_host_list.pop(0)
-
-        if (not self._is_friend_host(game) or
-           not self.announce_games or
-           game.state != GameState.OPEN):
-            return
-        self._announce(game, "hosting")
-
-    def _announce_replay(self, game):
-        if not self._is_friend_host(game) or not self.announce_replays:
-            return
-        self._announce(game, "playing live")
-
-    def _announce(self, game, activity):
-        url = game.url(game.host_player.id).toString()
-        url_color = self._colors.getColor("url")
-        mapname = maps.getDisplayName(game.mapname)
-        fmt = 'is {} {}<a style="color:{}" href="{}">{}</a> (on {})'
-        if game.featured_mod == "faf":
-            modname = ""
+        self._delayed_event_list = []
+
+    def _friend_event(self, player):
+        if self._client.init_ready:
+            self._friend_announce(player)
+        else:
+            self._delayed_event_list.append(player)
+
+    def delayed_friend_events(self):
+        while len(self._delayed_event_list) > 0:
+            player = self._delayed_event_list.pop(0)
+            self._friend_announce(player)
+
+    def _friend_announce(self, player):
```

Comment:
I think catching the 'event type' enum would make processing this much more readable. We can put processing each event type into a separate function, with some sanity checks to make sure things didn't change in the meantime.

Reply:
_friend_announce covers changes in the meantime.
but I guess .. I have to do what you say anyway .. else no cookie ... so why argue.

Reply:
I'm allergic to functions that don't fit on my 25x80 vga console :)

Reply:
That's what I suspected. (cause all your lines before 80)

Reply:
there is only one line over 80  (and that could be easy broken), and I do now typos and miss stuff 
.. I'm adapting ;-)